Windows Hotplug Support and Driver Name API

ids-imaging opened the pull request:
This adds hotplug support for the two windows backends, as well as an API to get the name of the driver that is used for a device (this is only relevant for windows).

This implementation uses the already existing enumeration functionality which means that this could be made more efficient in the future.

Feedback is very much appreciated!

#86

changyp6 commented:
The patch cannot be built in linux MinGW environment.
Error is Windows.h no such file or directory.

The reason is that in Windows system, "Windows" and "windows" are the same, but in Linux system, they are different.

ids-imaging commented:
Locally compiled with MSVC 2010, 2013, 2015, 2017, 2019 and msys2/mingw. Once all feedback is incorporated the changes will be squashed.

changyp6 commented:
```
libusb/os/windows_winusb.c:28:10: fatal error: Windows.h: No such file or directory
 #include <Windows.h>
          ^~~~~~~~~~~
```
Some capital letters still there.
```
libusb/os/windows_winusb.c:3484:12: warning: unknown conversion type character 'z' in format [-Wformat=]
   usbi_dbg("invalid size (%zu)", *size);
            ^~~~~~~~~~~~~~~~~~~~
libusb/os/windows_winusb.c:3563:12: warning: unknown conversion type character 'z' in format [-Wformat=]
   usbi_dbg("invalid size (%zu)", *size);
            ^~~~~~~~~~~~~~~~~~~~
```
These are just some ugly warnings.
`%zu` can only work in POSIX system, in Windows, it should be `%Iz`

changyp6 commented:
Sorry, my mistake.
It should be `%Iu` not `%Iz`. `%Iz` is a typo.

Please also remove `usbi_dbg("")` empty log from files. these lines are causing ugly warnings when compiling in MinGW.

And some other warnings saying that `static void windows_destroy_event_monitor() is NOT function prototype`,
The correct one should be `static void windows_destroy_event_monitor(void)`
and `static int windows_start_event_monitor(void)`
and `static void windows_destroy_event_window(void)`
and `static int windows_event_window_setup(void)`

It seems that `void` should be added to the parameter list, when there's no parameter of a function.

changyp6 commented:
After all these modifications, I can successfully compile libusb with hotplug support.
And I have tested, hotplug works as expected.
The registered device with specific VID can be reported.

But there's another issue.
`libusb_control_transfer()` doesn't seem to work.
but `libusb_bulk_transfer()` works.

ids-imaging commented:
I don't think that the empty debug logs should be removed by this pull request as I'm not sure that they were added to implicitly print the function call flow.

Which backend are you using and has the controll transfer been working before? At the moment I'm unsure what could cause that.

changyp6 commented:
I'm using libusb-win32.
The backend should be libusb0.sys

LudovicRousseau commented on libusb/core.c:
```diff
@@ -1151,11 +1167,31 @@ int API_EXPORTED libusb_get_max_iso_packet_size(libusb_device *dev,
  */
 DEFAULT_VISIBILITY
-libusb_device * LIBUSB_CALL libusb_ref_device(libusb_device *dev)
+libusb_device *LIBUSB_CALL libusb_ref_device(libusb_device *dev)
```
Why this change?

ids-imaging replied:
Because I use clang format with the linux kernel style as that is what most code in libusb seems to use. If this whitespace bothers you I an revert it.

LudovicRousseau commented on libusb/os/darwin_usb.c:
```diff
@@ -2247,19 +2247,29 @@ static int darwin_free_streams (struct libusb_device_handle *dev_handle, unsigne
         .get_device_descriptor = darwin_get_device_descriptor,
         .get_active_config_descriptor = darwin_get_active_config_descriptor,
         .get_config_descriptor = darwin_get_config_descriptor,
-        .hotplug_poll = darwin_hotplug_poll,
```
Why moving this line?

ids-imaging replied:
Because other platfoms have it there.

LudovicRousseau commented on libusb/os/haiku_usb_raw.cpp:
```diff
@@ -209,6 +209,15 @@ haiku_clock_gettime(int clkid, struct timespec *tp)
+	}
+	return 1;
+}
+
```
Maybe it is possible to have a `default_get_device_driver() ` to not have the define the same function in 6 different backends. Only the windows backend uses a different implementation.

ids-imaging replied:
I thought about it, but where would I put that function?

LudovicRousseau replied:
Maybe in `core.c` since it is common code?

I am not sure the function should return an empty string on Unix systems. Maybe returning `LIBUSB_ERROR_NOT_SUPPORTED` is better.
In fact I don't know why this new API is needed. Is it mandatory for hotplug?
Maybe you should split the patch and move the addition of `libusb_get_driver()` in a different Pull Request.

ids-imaging replied:
The addition to the API is important for us, not hotplugging.

**Why?**
Because some driver result in blue screens and that is something that has to be prevented. So this API is usefull for all windows users that have to make sure that a sane driver is used as backend. I thought I added that to the commit message but I must have forgotten it.

**API**
Ok that can be done.

LudovicRousseau commented on libusb/os/sunos_usb.c:
```diff
@@ -1666,6 +1675,7 @@ const struct usbi_os_backend usbi_backend = {
         .hotplug_poll = NULL,
         .open = sunos_open,
         .close = sunos_close,
+        .get_device_driver =sunos_get_device_driver,
```
missing space character after the "="

ids-imaging commented:
I found a deadlock introduced with the datarace commit, which would explain why the control transfer failed. I'll fix that tomorrow.

changyp6 commented:
> I found a deadlock introduced with the datarace commit, which would explain why the control transfer failed. I'll fix that tomorrow.

Thank you very much!

LudovicRousseau commented:
I propose you to separate the different changes in different PR.
For example your commits https://github.com/libusb/libusb/pull/617/commits/5c1fbde5dc8176f111f38aee0a020aa2708ff67f and https://github.com/libusb/libusb/pull/617/commits/6b028baf1eb78193cbc6b9daf1ff6b589b2bac71 should be in 2 different PR.

1 PR per issue solved.

changyp6 commented:
I've just cross-compiled this patch in MinGW environment, there are some warnings:
```
../../libusb/os/windows_winusb.c:3486:12: warning: unknown conversion type character 'l' in format [-Wformat=]
   usbi_dbg("invalid size (%Ilu)", *size);
            ^~~~~~~~~~~~~~~~~~~~~
../../libusb/libusbi.h:238:71: note: in definition of macro '_usbi_log'
 #define _usbi_log(ctx, level, ...) usbi_log(ctx, level, __FUNCTION__, __VA_ARGS__)
                                                                       ^~~~~~~~~~~
../../libusb/os/windows_winusb.c:3486:3: note: in expansion of macro 'usbi_dbg'
   usbi_dbg("invalid size (%Ilu)", *size);
   ^~~~~~~~
../../libusb/os/windows_winusb.c:3486:12: warning: too many arguments for format [-Wformat-extra-args]
   usbi_dbg("invalid size (%Ilu)", *size);
            ^~~~~~~~~~~~~~~~~~~~~
../../libusb/libusbi.h:238:71: note: in definition of macro '_usbi_log'
 #define _usbi_log(ctx, level, ...) usbi_log(ctx, level, __FUNCTION__, __VA_ARGS__)
                                                                       ^~~~~~~~~~~
../../libusb/os/windows_winusb.c:3486:3: note: in expansion of macro 'usbi_dbg'
   usbi_dbg("invalid size (%Ilu)", *size);
   ^~~~~~~~
../../libusb/os/windows_winusb.c: In function '_hid_set_report':
../../libusb/os/windows_winusb.c:3565:12: warning: unknown conversion type character 'l' in format [-Wformat=]
   usbi_dbg("invalid size (%Ilu)", *size);
            ^~~~~~~~~~~~~~~~~~~~~
```

`%Ilu` should be changed to `%Iu`

And
```
../libusb/os/poll_windows.c:63:13: warning: function declaration isn't a prototype [-Wstrict-prototypes]
 static void usbi_dec_fd_table()
             ^~~~~~~~~~~~~~~~~
```
`static void usbi_dec_fd_table()` should be `static void usbi_dec_fd_table(void)`

And
```
../../libusb/os/windows_usbdk.c:31:5: warning: "_MSC_VER" is not defined, evaluates to 0 [-Wundef]
 #if _MSC_VER >= 1800
     ^~~~~~~~
../../libusb/os/windows_usbdk.c:33:7: warning: "_MSC_VER" is not defined, evaluates to 0 [-Wundef]
 #elif _MSC_VER >= 1700
       ^~~~~~~~
```

I think these codes should be protected by 
```
#ifdef _MSC_VER
#endif
```

I think following code is sufficient enough
```
#ifdef _MSC_VER
 #if _MSC_VER >= 1700
 #include <usbiodef.h>
 #endif
#endif
```

Because `_MSC_VER >= 1700` has already covered `_MSC_VER >= 1800`

Anyhow, this pull request has made my 2018/2019 new year's wish come true!
Great job!!

ids-imaging commented:
> ...
> `%Ilu` should be changed to `%Iu`

That was an ugly issue - depending on the system the `l` was missing or not.

> ...
> `static void usbi_dec_fd_table()` should be `static void usbi_dec_fd_table(void)`

Not from this pull request, but I'll fix it.

> ...
> Because `_MSC_VER >= 1700` has already covered `_MSC_VER >= 1800`

Actually that block can be removed completely, since the fallback from `windows_nt_common.h:123` is good enough. Good catch!

> Anyhow, this pull request has made my 2018/2019 new year's wish come true!
> Great job!!

Thanks for testing and feedback!

changyp6 commented:
> > ...
> > `%Ilu` should be changed to `%Iu`
> 
> That was an ugly issue - depending on the system the `l` was missing or not.
> 
In my build environment, Fedora 30 with MinGW, `%Iu` is sufficient enough for both i686 and x86_64. Because `%Iu` is just like `%zu`. `I` works as `z`

I don't know whether other compiler complains about it or not.

ids-imaging commented on libusb/os/linux_usbfs.c:
```diff
@@ -2857,6 +2857,7 @@ const struct usbi_os_backend usbi_backend = {
 	.open = op_open,
 	.close = op_close,
+	.get_device_driver = NULL,
 	.get_configuration = op_get_configuration,
```
Should I also remove line 2850, for consistency purposes? Or do you want the smallest change sets possible?

LudovicRousseau replied:
Keep the patch as small as possible.
If you want to do other changes then open another Pull Request.

ids-imaging commented on libusb/os/openbsd_usb.c:
```diff
@@ -95,6 +95,7 @@ const struct usbi_os_backend usbi_backend = {
 	.close = obsd_close,
 
+	.get_device_driver = NULL,
 	.get_device_descriptor = obsd_get_device_descriptor,
```
What about haiku and darwin?

ids-imaging commented on libusb/os/netbsd_usb.c:
```diff
@@ -97,6 +97,7 @@ const struct usbi_os_backend usbi_backend = {
 	netbsd_close,
 
+	NULL,				/* netbsd_get_device_driver */,
 	netbsd_get_device_descriptor,
```
Should I change that? With redundant NULLs, so it's easier to work with in the future or without?

LudovicRousseau replied:
netbsd_usb.c does not use the `.field = ` notation so the extra NULL in the correct place is mandatory here.

ids-imaging commented on libusb/os/sunos_usb.c:
```diff
@@ -1666,6 +1666,7 @@ const struct usbi_os_backend usbi_backend = {
         .open = sunos_open,
         .close = sunos_close,
+        .get_device_driver = NULL,
         .get_configuration = sunos_get_configuration,
```
Should I also remove line 1666, 1678, 1679 and 1686 (wrong indent too?) for consistency purposes? Or do you want the smallest change sets possible?

ids-imaging commented:
Am I supposed to close the requests for change? Should I rebase/squash the changes so this can be merged?

LudovicRousseau commented:
You should create different PR for your different/unrelated changes.
See my previous comment https://github.com/libusb/libusb/pull/617#issuecomment-525346327

It is much more easy to review small PR.

hjelmn commented:
Yeah. Separate out each feature or bug fix into its own PR then squash the PRs down to as small a number of commits as possible. Try to make sure that the code still builds between commits so as not to break bisect.

hjelmn commented on libusb/core.c:
```diff
@@ -626,7 +626,7 @@ static struct discovered_devs *discovered_devs_alloc(void)
 }
 
-static void discovered_devs_free(struct discovered_devs *discdevs)
+void discovered_devs_free(struct discovered_devs *discdevs)
```
This needs to be renamed to ```usbi_discovered_devs_free``` to be consistent with the internal API naming.

hjelmn commented on libusb/core.c:
```diff
@@ -614,7 +614,7 @@ libusb_free_device_list(list, 1);
 
-static struct discovered_devs *discovered_devs_alloc(void)
+struct discovered_devs *discovered_devs_alloc(void)
 {
```
This needs to be renamed to ```usbi_discovered_devs_alloc``` to be consistent with the internal API naming.

hjelmn replied:
thanks!

hjelmn commented:
Thanks for working on this. It is getting very close.

ids-imaging commented:
I have incorporated the feedback and squashed all commits.

I see however no reason why there should be 6 PRs, 3 of those would be commits with less than 5 lines changed and one with about 40 – each PR containing one commit. The other two PRs still contain only one and two commits, though they are longer. They are easily reviewed in that form, and since no more feedback is expected, it makes even less sense to split them as context would be lost now. Moreover does this project use a merge strategy where no merge commits are made, meaning that the existence of more PRs and  their branches is not recorded in the git history, making them useless. 
Please explain why I should waste my time splitting this change set.

hjelmn commented on libusb/os/poll_windows.c:
```diff
@@ -60,7 +60,7 @@ static size_t fd_count;
 
-static void usbi_dec_fd_table()
+static void usbi_dec_fd_table(void)
 {
```
This are semantically equivalent. No problem with the change though.

hjelmn commented:
I can't comment on any of the backed changes. Otherwise LGTM.

ids-imaging commented:
There was a race condition between windows events and context initialization / scan of already present devices. This should be fixed with the last commit - on request I'll squash that commit.

Please also considder #635 and #634 for 1.0.24

ids-imaging commented:
Our QA is finding a troubeling amount of race condition - partially caused by this PR. I'll be closing this PR as well as #634 and #635 as result of that. The progress can be viewed and testet in our fork.

hjelmn commented:
@ids-imaging When do you expect to open a new PR? I wanted to get this into v1.0.24. I am willing to hold up the release until mid-November if you think it will be ready by then.

DiscreteLogarithm commented:
@ids-imaging Thank you for your work. Can you give us an update on the status of windows hotplug support, please. Also how serious are the race conditions? Can we use the library with locking on our side and if yes for which functions we need to do that? Thanks a lot in advance.

dickens commented:
Thank you for the effort on this patch.  Windows hotplug is a long-awaited feature.  Unfortunately as  you discovered this implementation is not going to work.  There are also some things that are not completely addressed in this approach.

With that said, I'll be working on rebasing my previous hotplug work onto master.

mcuee commented:
Great to see Chris to be working on this again. This feature is really the major gap for the Windows backend. And Windows backend is really more complicated than the others.

liziqi-matrixrealm commented:
libusb_open or libusb_open_device_with_vid_pid  failed and return -5 LIBUSB_ERROR_NOT_FOUND after hotplug libusb_hotplug_register.

liziqi-matrixrealm commented:
Report a bug In windows platform.  There is a race when hotplug in then use winusb_get_device_list immediately. The USB device's emulation or some initialization has not finished so return -5 LIBUSB_ERROR_NOT_FOUND. We can add a short sleep before winusb_get_device_list to solve it.
